server/save.go

```diff
@@ -1,83 +0,0 @@
-package main
-
-import (
-	"encoding/json"
-	"time"
-
-	"github.com/pkg/errors"
-)
-
-func (p *Plugin) retreiveData() error {
-	j, err := p.API.KVGet("analytics")
-	if err != nil {
-		return errors.Wrap(err, "failed to get analytics from kv")
-	}
-	p.currentAnalytic = NewAnalytic()
-	if err := json.Unmarshal(j, p.currentAnalytic); err != nil {
-		p.API.LogError("failed to unmarshal analytics from kv use new one", "err", err.Error())
-		p.currentAnalytic = NewAnalytic()
-	}
-	return nil
-}
-
-func (p *Plugin) startCronSaver(poison chan bool) {
-	go func() {
-		for {
-			select {
-			case <-time.After(1 + time.Minute):
-				p.save()
-			case <-poison:
-				p.save()
-				return
-			}
-		}
-	}()
-}
-
-func (p *Plugin) save() {
-	p.currentAnalytic.RLock()
-	defer p.currentAnalytic.RUnlock()
-	j, err := json.Marshal(p.currentAnalytic)
-	if err != nil {
-		p.API.LogError("can't marshal internal analytics data")
-	}
-	if err := p.API.KVSet("analytics", j); err != nil {
-		p.API.LogError("can't save analytics data")
-	}
-}
-
-func (p *Plugin) allSessions() ([]*Analytic, error) {
-	allAnalytics := make([]*Analytic, 0)
-
-	j, err := p.API.KVGet("allAnalytics")
-	if err != nil {
-		p.API.LogError("can't get allAnalytics", "err", err.Error())
-		return nil, err
-	}
-
-	if err := json.Unmarshal(j, &allAnalytics); err != nil {
-		p.API.LogError("failed to unmarshal analytics from kv use new one", "err", err.Error())
-		return nil, err
-	}
-
-	return allAnalytics, nil
-}
-
-func (p *Plugin) newSession() {
-	p.currentAnalytic.WLock()
-	defer p.currentAnalytic.WUnlock()
-
-	allAnalytics, err := p.allSessions()
-	if err != nil {
-		p.API.LogWarn("can't get all sessions", "err", err.Error())
-	}
-
-	j2, err2 := json.Marshal(append(allAnalytics, p.currentAnalytic.Close()))
-	if err2 != nil {
-		p.API.LogWarn("can't marshal internal analytics data")
-	}
-	if err := p.API.KVSet("allAnalytics", j2); err != nil {
```

Comment:
I don't understand why you use two key. What is the difference between `allAnalytics` and `analytics`?

Reply:
`analytics` is the current session of analytics, `allAnalytics` is an array with all sessions. `analytics` is loaded when plugin start, stored every minute in case of failure. The second is loaded when a report is requested to display all data : 

![selection_186](https://user-images.githubusercontent.com/1492516/53492963-73948a80-3a9a-11e9-9e5c-e76aff4ecc92.png)